pkg/vcdsdk/gateway.go

```diff
@@ -1987,3 +1986,131 @@ func (gm *GatewayManager) FetchIpSpacesBackingGateway(ctx context.Context) ([]st
+	}
+
+	return result[0].ID, result[0].Value, nil
+}
+
+func (gm *GatewayManager) FindIpAllocationByIp(ipSpace *govcd.IpSpace, allocatedIp string) (*govcd.IpSpaceIpAllocation, error) {
+	queryParams := url.Values{}
+	queryParams.Set("filter", fmt.Sprintf("value==%s", allocatedIp))
+
+	allocations, err := ipSpace.GetAllIpSpaceAllocations(types.IpSpaceIpAllocationTypeFloatingIp, queryParams)
+	if err != nil {
+		return nil, fmt.Errorf("unable to find allocation in Ip Space [%s], correspnding to Ip [%s]. error [%v]", ipSpace.IpSpace.Name, allocatedIp, err)
+	}
+
+	// If no allocations were made on the Ip Space with the specified Ip, it is valid to return no result with no error
+	// This use case is important for us, hence skipping using GetIpSpaceAllocationByTypeAndValue
+	if allocations == nil || len(allocations) == 0 {
+		return nil, nil
+	}
+
+	if len(allocations) == 1 {
+		return allocations[0], nil
+	}
+
+	// Ideally we should never reach here, since VCD shouldn't make multiple allocations for the same IP
+	return nil, fmt.Errorf("found multiple allocations in Ip Space [%s] for Ip [%s]", ipSpace.IpSpace.Name, allocatedIp)
+}
+
+// FindIpAllocationByMarker We are adding info in the description of IpAllocation
+// to record that particular allocation was made by a CSE cluster. This method finds an
+// allocation corresponding to the previously stored description
+func (gm *GatewayManager) FindIpAllocationByMarker(ipSpace *govcd.IpSpace, marker string) (*govcd.IpSpaceIpAllocation, error) {
+	queryParams := url.Values{}
+	queryParams.Set("filter", fmt.Sprintf("usageState==%s", types.IpSpaceIpAllocationUsedManual))
+
+	allocations, err := ipSpace.GetAllIpSpaceAllocations(types.IpSpaceIpAllocationTypeFloatingIp, queryParams)
+	if err != nil {
+		return nil, fmt.Errorf("unable to find allocation in Ip Space [%s], correspnding to marker [%s]. error [%v]", ipSpace.IpSpace.Name, marker, err)
+	}
+
+	// If no allocations were made on the Ip Space, it is valid to return no result without error
+	if allocations == nil || len(allocations) == 0 {
+		return nil, nil
+	}
+
+	for _, allocation := range allocations {
+		if allocation.IpSpaceIpAllocation.Description == marker {
+			return allocation, nil
+		}
+	}
+
+	// If no allocation with the marker is found it is ok to return back with no result and no error
+	return nil, nil
+}
+
+// MarkIpAsUsed We are adding info in the description  of IpAllocation
+// to track that particular allocation was made by CSE cluster. This method updates
+// the state and description of an allocation to mark the cluster which owns the allocation.
+func (gm *GatewayManager) MarkIpAsUsed(ipAllocation *govcd.IpSpaceIpAllocation, marker string) (*govcd.IpSpaceIpAllocation, error) {
+	updateConfig := ipAllocation.IpSpaceIpAllocation
+	updateConfig.UsageState = types.IpSpaceIpAllocationUsedManual
+	updateConfig.Description = marker
```

Comment:
It will be helpful to have a comment here saying something like:
// Note: we do not check the `Description` here because there are no concurrency guarantees even
// if a check is made. We depend on Kubernetes framework to ensure that a single worker is used to
// run the EnsureLoadBalancerAvailable task.

Reply:
Will add this comment in the piece of code which calls this function and AllocateIp